ethcontract-common/src/artifact.rs

```diff
@@ -188,6 +189,18 @@ impl Deref for ContractMut<'_> {
     }
 }
 
+impl Drop for ContractMut<'_> {
+    fn drop(&mut self) {
+        // The ABI might have gotten mutated while this guard was alive.
+        // Since we compute pre-compute and cache a few values based on the ABI
+        // as a performance optimization we need to recompute those cached values
+        // with the new ABI once the user is done updating the mutable contract.
+        let abi = self.0.interface.abi.clone();
+        let interface = Interface::from(abi);
+        *Arc::make_mut(&mut self.0.interface) = interface;
```

Comment:
Can this panic if there is more than one `Arc` pointer to the data? Asking cuz my Rust is rusty :P

Reply:
No, this function is actually pretty cool. If there are other Arcs this will clone the current content of the Arc and make it point to the new allocation. If it's the only `Arc` it will mutate the content in place. It's slightly wonky because you could clone the original Arc, update the artifact and expect that your cloned Arc would see the updates but this is not the case. I think this is fine, though since fundamentally this behavior was already present before factoring in borrowing rules.

Reply:
Oh, cool - so it has CoW (copy on write) semantics. Very cool, and the name is a great coincidence 😅.

> I think this is fine, though since fundamentally this behavior was already present before factoring in borrowing rules.

I would even argue this is desired to preserve the existing mutation semantics.